### PR TITLE
feat: Process message and schema annotations

### PR DESCRIPTION
**Pull request by lorenz-scalable**
### What
This is the first step to process AsyncAPI annotations to support meta-documentation. We register an annotation provider bean that scans the classpath for message and schema annotations. Those annotations will be processed considering the context (annotated class) and they will be translated to AsyncAPI components.

### Why
- this will allow users to annotate their event-based microservices and generate AsyncAPI documentation automatically.

### How
- AnnotationProvider: uses the annotation scanner and processors to generate components
- AnnotationScanner:  scans the classpath for specified annotations
- AnnotationProcessor: processes annotation with the specified context (e.g. annotated class)

### CODE REVIEW

File: kotlin-asyncapi-annotation/src/main/kotlin/org/openfolder/kotlinasyncapi/annotation/channel/MessageExample.kt [hunk 1]
```diff
@@ -1,10 +1,8 @@
 package org.openfolder.kotlinasyncapi.annotation.channel
 
-import org.openfolder.kotlinasyncapi.annotation.Schema
-
 annotation class MessageExample(
     val default: Boolean = false,
-    val headers: Schema = Schema(default = true),
+    val headers: String = "",
```
Comment (drazengrabovac) [inline kotlin-asyncapi-annotation/src/main/kotlin/org/openfolder/kotlinasyncapi/annotation/channel/MessageExample.kt:5]: Why was this changed to String?

File: kotlin-asyncapi-annotation/src/main/kotlin/org/openfolder/kotlinasyncapi/annotation/Schema.kt [hunk 2]
```diff
@@ -2,34 +2,13 @@ package org.openfolder.kotlinasyncapi.annotation
...
-    val additionalProperties: String = "",
-    val allOf: Array<KClass<*>> = [],
-    val anyOf: Array<KClass<*>> = [],
-    val oneOf: Array<KClass<*>> = [],
-    val not: KClass<*> = Void::class,
-    val format: String = "",
-    val discriminator: String = "",
-    val externalDocs: ExternalDocumentation = ExternalDocumentation(default = true, url = ""),
-    val deprecated: Boolean = false
+    val implementation: KClass<*> = Void::class
```
Comment (drazengrabovac) [inline kotlin-asyncapi-annotation/src/main/kotlin/org/openfolder/kotlinasyncapi/annotation/Schema.kt:13]: Why did we remove all of the schema fields? How is the `implementation` field replacing this?

File: kotlin-asyncapi-core/src/test/kotlin/org/openfolder/kotlinasyncapi/model/AsyncApiIntegrationTest.kt [hunk 3]
```diff
@@ -79,10 +79,8 @@ internal class AsyncApiIntegrationTest {
                                 }
                             }
                         }
-                        bindings {
-                            reference("http") {
-                                ref("#/components/messageBindings/streamingHeaders")
-                            }
+                        bindingsRef {
+                            ref("#/components/messageBindings/streamingHeaders")
```
Comment (drazengrabovac) [inline kotlin-asyncapi-core/src/test/kotlin/org/openfolder/kotlinasyncapi/model/AsyncApiIntegrationTest.kt:83]: Just for my understanding, we will not depend on OpenApi specs anymore for the components part right?

File: kotlin-asyncapi-spring-web/src/main/kotlin/org/openfolder/kotlinasyncapi/springweb/context/annotation/processor/MappingUtils.kt [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+
+internal fun MessageBinding.Kafka.toKafka(): org.openfolder.kotlinasyncapi.model.channel.MessageBinding.Kafka =
+    org.openfolder.kotlinasyncapi.model.channel.MessageBinding.Kafka().apply {
+        bindingVersion = this@toKafka.bindingVersion.takeIf { it.isNotEmpty() }
+        key = this@toKafka.key.takeUnless { it.default }?.implementation?.let {
+            Reference().apply { ref("#/components/schemas/${it.simpleName}") }
+        }
+    }
+
+internal fun MessageBinding.AnypointMQ.toAnypointMQ(): org.openfolder.kotlinasyncapi.model.channel.MessageBinding.AnypointMQ =
```
Comment (drazengrabovac) [inline kotlin-asyncapi-spring-web/src/main/kotlin/org/openfolder/kotlinasyncapi/springweb/context/annotation/processor/MappingUtils.kt:134]: Where does SQS/SNS fit in here? Is it this Anypoint?

**Review by drazengrabovac — APPROVED** [review]
Great ground work! I just a have a couple of small questions.

File: kotlin-asyncapi-annotation/src/main/kotlin/org/openfolder/kotlinasyncapi/annotation/channel/MessageExample.kt (hunk 1 above)
Comment (lorenz-scalable) [reply kotlin-asyncapi-annotation/src/main/kotlin/org/openfolder/kotlinasyncapi/annotation/channel/MessageExample.kt:5]: I initially made a mistake. this field holds an object and not a schema. see the integration test for clarification.

File: kotlin-asyncapi-annotation/src/main/kotlin/org/openfolder/kotlinasyncapi/annotation/Schema.kt (hunk 2 above)
Comment (lorenz-scalable) [reply kotlin-asyncapi-annotation/src/main/kotlin/org/openfolder/kotlinasyncapi/annotation/Schema.kt:13]: we currently would ignore all other fields because we only use the implementation field to generate the schema object from it. we would need to implement an extension logic to also include the other field, which is not feasible for now.

File: kotlin-asyncapi-core/src/test/kotlin/org/openfolder/kotlinasyncapi/model/AsyncApiIntegrationTest.kt (hunk 3 above)
Comment (lorenz-scalable) [reply kotlin-asyncapi-core/src/test/kotlin/org/openfolder/kotlinasyncapi/model/AsyncApiIntegrationTest.kt:83]: where did we depend on openapi before?

File: kotlin-asyncapi-spring-web/src/main/kotlin/org/openfolder/kotlinasyncapi/springweb/context/annotation/processor/MappingUtils.kt (hunk 4 above)
Comment (lorenz-scalable) [reply kotlin-asyncapi-spring-web/src/main/kotlin/org/openfolder/kotlinasyncapi/springweb/context/annotation/processor/MappingUtils.kt:134]: sqs and sns also have message bindings but they are not implemented by AsyncAPI for now https://github.com/asyncapi/bindings/blob/master/sqs/README.md. that is why we dont have a mapping here.

**Review by mohaqeq — APPROVED** [review]

**Comment by mohaqeq** [comment]
I didn't run the tests locally and hope they all work ;)
